File: health_sync/providers/hevy.py
from __future__ import annotations

from datetime import timedelta

import requests

from ..config import LoadedConfig
from ..db import HealthSyncDb
from ..util import dt_to_iso_z, iso_to_dt, request_json, sha256_hex, utc_now_iso

HEVY_PROVIDER = "hevy"
HEVY_BASE_DEFAULT = "https://api.hevyapp.com"


def _hevy_api_key(cfg: LoadedConfig) -> str:
    if cfg.config.hevy.api_key:
        return cfg.config.hevy.api_key
    raise RuntimeError(
        "Missing Hevy API key. Set `hevy.api_key` in your config file "
        f"({cfg.path}). You can get it from https://hevy.com/settings?developer"
    )


def _hevy_headers(api_key: str) -> dict[str, str]:
    return {"api-key": api_key, "User-Agent": "health-sync/0.1 (+local sqlite cache)", "Accept": "application/json"}


def _iso_max(a: str | None, b: str | None) -> str | None:
    if not a:
        return b
    if not b:
        return a
    try:
        return a if iso_to_dt(a) >= iso_to_dt(b) else b
    except Exception:  # noqa: BLE001
        return a


def _page_size(cfg: LoadedConfig) -> int:
    return max(1, min(10, int(cfg.config.hevy.page_size)))


def hevy_sync(db: HealthSyncDb, cfg: LoadedConfig) -> None:
    api_key = _hevy_api_key(cfg)
    sess = requests.Session()
    base_url = cfg.config.hevy.base_url or HEVY_BASE_DEFAULT
    state = db.get_sync_state(provider=HEVY_PROVIDER, resource="workouts")
    watermark = state.watermark if state else None
    page_size = _page_size(cfg)
    overlap_s = int(cfg.config.hevy.overlap_seconds)

    print("Syncing Hevy...")
    if not watermark:
        with db.sync_run(provider=HEVY_PROVIDER, resource="workouts") as run:
            with db.transaction():
                page, max_updated = 1, None
                while True:
                    j = request_json(
                        sess,
                        "GET",
                        f"{base_url}/v1/workouts",
                        headers=_hevy_headers(api_key),
                        params={"page": page, "pageSize": page_size},
                    )
                    for w in j.get("workouts") or []:
                        if not isinstance(w, dict):
                            continue
                        updated_at = w.get("updated_at") or w.get("created_at")
                        max_updated = _iso_max(max_updated, updated_at)
                        run.add_upsert(
                            db.upsert_record(
                                provider=HEVY_PROVIDER,
                                resource="workouts",
                                record_id=str(w.get("id") or sha256_hex(str(w))),
                                payload=w,
                                start_time=w.get("start_time"),
                                end_time=w.get("end_time"),
                                source_updated_at=updated_at,
                            )
                        )
                    if isinstance(j.get("page_count"), int) and page < j["page_count"]:
                        page += 1
                        continue
                    break
                db.set_sync_state(provider=HEVY_PROVIDER, resource="workouts", watermark=max_updated or utc_now_iso())
        print("Hevy initial backfill complete.")
        return

    try:
        since = dt_to_iso_z(iso_to_dt(watermark) - timedelta(seconds=overlap_s))
    except Exception:  # noqa: BLE001
        since = cfg.config.hevy.since

    with db.sync_run(provider=HEVY_PROVIDER, resource="workouts") as run_workouts:
        with db.sync_run(provider=HEVY_PROVIDER, resource="workout_events") as run_events:
            with db.transaction():
                page, max_event_time = 1, watermark
                while True:
                    j = request_json(
                        sess,
                        "GET",
                        f"{base_url}/v1/workouts/events",
                        headers=_hevy_headers(api_key),
                        params={"since": since, "page": page, "pageSize": page_size},
                    )
                    for ev in j.get("events") or []:
                        if not isinstance(ev, dict):
                            continue
                        ev_type = ev.get("type")
                        if ev_type == "updated" and isinstance(ev.get("workout"), dict):
                            w = ev["workout"]
                            rid = str(w.get("id") or sha256_hex(str(w)))
                            updated_at = w.get("updated_at") or w.get("created_at")
                            max_event_time = _iso_max(max_event_time, updated_at)
                            run_workouts.add_upsert(
                                db.upsert_record(
                                    provider=HEVY_PROVIDER,
                                    resource="workouts",
                                    record_id=rid,
                                    payload=w,
                                    start_time=w.get("start_time"),
                                    end_time=w.get("end_time"),
                                    source_updated_at=updated_at,
                                )
                            )
                            run_events.add_upsert(
                                db.upsert_record(
                                    provider=HEVY_PROVIDER,
                                    resource="workout_events",
                                    record_id=f"updated:{rid}:{updated_at or utc_now_iso()}",
                                    payload=ev,
                                    start_time=updated_at,
                                    end_time=None,
                                    source_updated_at=updated_at,
                                )
                            )
                        elif ev_type == "deleted":
                            rid, deleted_at = str(ev.get("id") or ""), ev.get("deleted_at")
                            max_event_time = _iso_max(max_event_time, deleted_at)
                            if rid:
                                run_workouts.add_delete(
                                    db.delete_record(provider=HEVY_PROVIDER, resource="workouts", record_id=rid)
                                )
                                run_events.add_upsert(
                                    db.upsert_record(
                                        provider=HEVY_PROVIDER,
                                        resource="workout_events",
                                        record_id=f"deleted:{rid}:{deleted_at or utc_now_iso()}",
                                        payload=ev,
                                        start_time=deleted_at,
                                        end_time=None,
                                        source_updated_at=deleted_at,
                                    )
                                )
                        else:
                            run_events.add_upsert(
                                db.upsert_record(
                                    provider=HEVY_PROVIDER,
                                    resource="workout_events",
                                    record_id=f"unknown:{sha256_hex(str(ev))}",
                                    payload=ev,
                                )
                            )
                    if isinstance(j.get("page_count"), int) and page < j["page_count"]:
                        page += 1
                        continue
                    break
                if max_event_time and max_event_time != watermark:
                    db.set_sync_state(provider=HEVY_PROVIDER, resource="workouts", watermark=max_event_time)

    print("Hevy sync complete.")

File: src/providers/hevy.js
import {
  dtToIsoZ,
  requestJson,
  sha256Hex,
  toEpochSeconds,
} from '../util.js';

function clampPageSize(rawValue) {
  const parsed = Number.parseInt(String(rawValue ?? 10), 10);
  if (!Number.isFinite(parsed)) {
    return 10;
  }
  return Math.max(1, Math.min(10, parsed));
}

function isoFromEpochSeconds(epochSeconds) {
  const value = Number(epochSeconds);
  if (!Number.isFinite(value)) {
    return null;
  }
  return dtToIsoZ(new Date(Math.floor(value) * 1000));
}

function hevyHeaders(apiKey) {
  return {
    'api-key': apiKey,
  };
}

function workoutRecordShape(workout) {
  return {
    recordId: workout?.id ? String(workout.id) : sha256Hex(JSON.stringify(workout)),
    startTime: workout?.start_time || null,
    endTime: workout?.end_time || null,
    sourceUpdatedAt: workout?.updated_at || workout?.created_at || null,
  };
}

function eventTimestamp(event) {
  return event?.updated_at || event?.deleted_at || event?.timestamp || event?.created_at || null;
}

async function hevyInitialSync(db, cfg, apiKey) {
  await db.syncRun('hevy', 'workouts', async () => {
    await db.transaction(async () => {
      const baseUrl = String(cfg.base_url || 'https://api.hevyapp.com').replace(/\/$/, '');
      const pageSize = clampPageSize(cfg.page_size);

      let page = 1;
      let maxUpdatedEpoch = null;

      while (true) {
        const response = await requestJson(`${baseUrl}/v1/workouts`, {
          headers: hevyHeaders(apiKey),
          params: {
            page,
            pageSize,
          },
        });

        const workouts = Array.isArray(response?.workouts)
          ? response.workouts
          : Array.isArray(response?.data)
            ? response.data
            : [];

        for (const workout of workouts) {
          const shape = workoutRecordShape(workout);
          db.upsertRecord({
            provider: 'hevy',
            resource: 'workouts',
            recordId: shape.recordId,
            startTime: shape.startTime,
            endTime: shape.endTime,
            sourceUpdatedAt: shape.sourceUpdatedAt,
            payload: workout,
          });

          const updatedEpoch = toEpochSeconds(shape.sourceUpdatedAt);
          if (updatedEpoch !== null) {
            maxUpdatedEpoch = maxUpdatedEpoch === null
              ? updatedEpoch
              : Math.max(maxUpdatedEpoch, updatedEpoch);
          }
        }

        const pageCount = Number.parseInt(String(response?.page_count ?? response?.pageCount ?? 0), 10) || 0;
        if (!workouts.length || pageCount <= 0 || page >= pageCount) {
          break;
        }
        page += 1;
      }

      const watermarkEpoch = maxUpdatedEpoch ?? Math.floor(Date.now() / 1000);
      db.setSyncState('hevy', 'workouts', {
        watermark: isoFromEpochSeconds(watermarkEpoch),
      });
    });
  });
}

async function hevyDeltaSync(db, cfg, apiKey) {
  const baseUrl = String(cfg.base_url || 'https://api.hevyapp.com').replace(/\/$/, '');
  const pageSize = clampPageSize(cfg.page_size);
  const parsedOverlap = Number.parseInt(String(cfg.overlap_seconds ?? 300), 10);
  const overlapSeconds = Number.isFinite(parsedOverlap) ? Math.max(0, parsedOverlap) : 300;

  const state = db.getSyncState('hevy', 'workouts');
  const watermarkEpoch = toEpochSeconds(state?.watermark);
  const fallbackSinceEpoch = toEpochSeconds(cfg.since || '1970-01-01T00:00:00Z') || 0;
  const sinceEpoch = watermarkEpoch === null
    ? fallbackSinceEpoch
    : Math.max(fallbackSinceEpoch, Math.max(0, watermarkEpoch - overlapSeconds));

  await db.syncRun('hevy', 'workouts', async () => {
    await db.syncRun('hevy', 'workout_events', async () => {
      await db.transaction(async () => {
        let page = 1;
        let maxEventEpoch = watermarkEpoch;

        while (true) {
          const response = await requestJson(`${baseUrl}/v1/workouts/events`, {
            headers: hevyHeaders(apiKey),
            params: {
              page,
              pageSize,
              since: isoFromEpochSeconds(sinceEpoch),
            },
          });

          const events = Array.isArray(response?.events)
            ? response.events
            : Array.isArray(response?.data)
              ? response.data
              : [];

          for (const event of events) {
            const type = String(event?.type || '').toLowerCase();
            const eventTs = eventTimestamp(event);
            const eventEpoch = toEpochSeconds(eventTs);
            if (eventEpoch !== null) {
              maxEventEpoch = maxEventEpoch === null
                ? eventEpoch
                : Math.max(maxEventEpoch, eventEpoch);
            }

            if (type === 'updated') {
              const workout = event?.workout && typeof event.workout === 'object'
                ? event.workout
                : event?.data && typeof event.data === 'object'
                  ? event.data
                  : null;

              const workoutId = workout?.id || event?.workout_id || event?.id;
              if (workout && workoutId) {
                const shape = workoutRecordShape(workout);
                db.upsertRecord({
                  provider: 'hevy',
                  resource: 'workouts',
                  recordId: String(workoutId),
                  startTime: shape.startTime,
                  endTime: shape.endTime,
                  sourceUpdatedAt: shape.sourceUpdatedAt,
                  payload: workout,
                }, { provider: 'hevy', resource: 'workouts' });
              }

              const auditId = `updated:${workoutId || 'unknown'}:${eventTs || 'na'}`;
              db.upsertRecord({
                provider: 'hevy',
                resource: 'workout_events',
                recordId: auditId,
                startTime: eventTs,
                endTime: null,
                sourceUpdatedAt: eventTs,
                payload: event,
              }, { provider: 'hevy', resource: 'workout_events' });
              continue;
            }

            if (type === 'deleted') {
              const workoutId = event?.workout_id || event?.id || null;
              if (workoutId) {
                db.deleteRecord('hevy', 'workouts', String(workoutId), { provider: 'hevy', resource: 'workouts' });
              }

              const deletedTs = event?.deleted_at || eventTs;
              const auditId = `deleted:${workoutId || 'unknown'}:${deletedTs || 'na'}`;
              db.upsertRecord({
                provider: 'hevy',
                resource: 'workout_events',
                recordId: auditId,
                startTime: deletedTs,
                endTime: null,
                sourceUpdatedAt: deletedTs,
                payload: event,
              }, { provider: 'hevy', resource: 'workout_events' });
              continue;
            }

            const unknownId = `unknown:${sha256Hex(JSON.stringify(event))}`;
            db.upsertRecord({
              provider: 'hevy',
              resource: 'workout_events',
              recordId: unknownId,
              startTime: eventTs,
              endTime: null,
              sourceUpdatedAt: eventTs,
              payload: event,
            }, { provider: 'hevy', resource: 'workout_events' });
          }

          const pageCount = Number.parseInt(String(response?.page_count ?? response?.pageCount ?? 0), 10) || 0;
          if (!events.length || pageCount <= 0 || page >= pageCount) {
            break;
          }
          page += 1;
        }

        if (maxEventEpoch !== null) {
          db.setSyncState('hevy', 'workouts', {
            watermark: isoFromEpochSeconds(maxEventEpoch),
          });
        }
      });
    });
  });
}

async function hevySync(db, config, helpers) {
  const cfg = helpers.configFor('hevy');
  const apiKey = helpers.requireStr('hevy', 'api_key', 'Missing [hevy].api_key');

  const state = db.getSyncState('hevy', 'workouts');
  if (!state?.watermark) {
    await hevyInitialSync(db, cfg, apiKey);
    return;
  }

  await hevyDeltaSync(db, cfg, apiKey);
}

const hevyProvider = {
  id: 'hevy',
  source: 'builtin',
  description: 'Hevy workouts and workout event deltas',
  supportsAuth: false,
  sync: hevySync,
};

export default hevyProvider;
